libroach, keys: emulate `DB::SyncWAL()` on Windows

BramGruneir opened the pull request:
In windows, rocksdb's `DB:SyncWAL()` function is not implemented due to fact
that `WinWritableFile` is not thread safe. To get around that, the only other
methods than can be used to ensure that a sync is triggered is to either flush
the memtables or perform a write with `WriteOptions.sync=true`. See
https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ for more details.
Please also see #17442 for more discussion on the topic.

So the solution in this case is to issue a write to a single store local key
with `sync=true`. In this commit, that key reserved to ensure that we never
accidentally use it for some other purpose.

This has been tested on Windows.

Fixes #17442.

BramGruneir committed:
libroach, keys: emulate `DB::SyncWAL()` on Windows

On Windows, rocksdb's `DB::SyncWAL()` function is not implemented due to fact
that `WinWritableFile` is not thread safe. To get around that, the only other
methods than can be used to ensure that a sync is triggered is to either flush
the memtables or perform a write with `WriteOptions.sync=true`. See
https://github.com/facebook/rocksdb/wiki/RocksDB-FAQ for more details.
Please also see #17442 for more discussion on the topic.

So the solution in this case is to issue a write to a single store local key
with `sync=true`. In this commit, that key reserved to ensure that we never
accidentally use it for some other purpose.

This has been tested on Windows.

Fixes #17442.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17735)
<!-- Reviewable:end -->

benesch commented:
I'm not qualified to evaluate the key encoding scheme, but the general approach <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>. Thanks for doing this, @BramGruneir!

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-:-KrlODJUQLaiCUDnPEME:bchxort)*
<!-- Sent from Reviewable.io -->

benesch commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-KrlV5_cXv7iJX0XPtfN:bubysiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
> ```cpp
> }
> 
> const std::string encodedLocalStoreSyncKey = EncodeKey(ToDBKey(kKeyLocalStoreSync));
> ```

`ToDBKey` parses a key that already has timestamp information encoded in it. To encode the key here you need to construct a `DBKey` and set its `key` field (leaving the timestamp fields zero). 

---

*[pkg/keys/keys.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlUJyJJHKUx0ULtYyq:-KrlUJyJJHKUx0ULtYyr:bsset5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/pkg/keys/keys.go#L62)):*
> ```Go
> }
> 
> // StoreSyncKey returns a store-local key used on for emulating syncwall on
> ```

s/syncwall/SyncWAL/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-KrlXdHcRLShc6DhaD9O:bcgdxo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`ToDBKey` parses a key that already has timestamp information encoded in it. To encode the key here you need to construct a `DBKey` and set its `key` field (leaving the timestamp fields zero). 
</blockquote></details>

What @bdarnell said is correct. Constructing the `DBKey` is mildly irritating. You could append the NUL byte to `kKeyLocalStoreSync`: `"\x01ssync\x00"` (be sure to increase the length by 1).

---

*[c-deps/libroach/db.cc, line 1698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlRm-PcFAWodJEvcn0:-KrlRm-PcFAWodJEvcn1:b6hsa92) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1698)):*
> ```cpp
> DBStatus DBSyncWAL(DBEngine* db) {
>   #ifdef _WIN32
>     // In windows, DB:SyncWAL() is not implemented due to fact that
> ```

C++ syntax is double colons: `s/DB:SyncWal/DB::SyncWal/g`

`s/In windows/On windows/g`

---

*[c-deps/libroach/db.cc, line 1712 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlWMxrlLQzNPopn0ur:-KrlWMxrlLQzNPopn0us:b-6xdw4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1712)):*
> ```cpp
>     return ToDBStatus(db->rep->Put(options,
>       encodedLocalStoreSyncKey,
>       ToSlice(DBSlice())));
> ```

`s/ToSlice(DBSlice())/""/g`

I think the above can be folded into a single line.

---

*[pkg/keys/constants.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlWxiprMW9XPs2vc4o:-KrlWxiprMW9XPs2vc4p:b-s5kxws) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/pkg/keys/constants.go#L99)):*
> ```Go
> 	// So in order to force a sync, we issue a write to a single store local key
> 	// with `sync=true`. localStoreSyncSuffix is this key and is used solely for
> 	// this purpose.
> ```

Rather than copying this comment, I'd reference the one in `db.cc`:

```
// Windows-specific sync key. See `db.cc:DBSyncWAL` for details.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-KrlkRlP1ZKV-XV7Iu19:bcur6xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What @bdarnell said is correct. Constructing the `DBKey` is mildly irritating. You could append the NUL byte to `kKeyLocalStoreSync`: `"\x01ssync\x00"` (be sure to increase the length by 1).
</blockquote></details>

Just to be clear, by appending the NUL byte, this would would set the timestamp fields to zero?  If so, that's cleaner and I'll go with that route.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-KrlkrzYHPupXUbhvHng:b8hqs2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Just to be clear, by appending the NUL byte, this would would set the timestamp fields to zero?  If so, that's cleaner and I'll go with that route.
</blockquote></details>

Yes, that's what `EncodeKey` would be doing. The NUL byte is indicating that there are 0 bytes of timestamp suffix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-Krlm2xKoD-xn2DhHm0K:bm2v6v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, that's what `EncodeKey` would be doing. The NUL byte is indicating that there are 0 bytes of timestamp suffix.
</blockquote></details>

Awesome.  Done.

---

*[c-deps/libroach/db.cc, line 1698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlRm-PcFAWodJEvcn0:-Krll09lTfSyJxk1-ptB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1698)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

C++ syntax is double colons: `s/DB:SyncWal/DB::SyncWal/g`

`s/In windows/On windows/g`
</blockquote></details>

Done.

---

*[c-deps/libroach/db.cc, line 1712 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlWMxrlLQzNPopn0ur:-KrllWJN62lLoHYWg0Ll:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1712)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/ToSlice(DBSlice())/""/g`

I think the above can be folded into a single line.
</blockquote></details>

Done.

---

*[pkg/keys/constants.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlWxiprMW9XPs2vc4o:-KrlldnaD5b_ykbvqAO5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/pkg/keys/constants.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than copying this comment, I'd reference the one in `db.cc`:

```
// Windows-specific sync key. See `db.cc:DBSyncWAL` for details.
```
</blockquote></details>

Done.

---

*[pkg/keys/keys.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlUJyJJHKUx0ULtYyq:-KrllvSuUbIXcY2MtFOt:b6aeqpn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/pkg/keys/keys.go#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/syncwall/SyncWAL/
</blockquote></details>

Bah, no matter how proofread all the comments, typos always sneak in.

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-KrlqhLc_eaESGTMtRe-:bd7wj9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Awesome.  Done.
</blockquote></details>

`encodedLocalStoreSyncKey == kKeyLocalStoreSync`. Or it should. 

---

*[c-deps/libroach/db.cc, line 1698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlRm-PcFAWodJEvcn0:-KrlqvKbFBvmw24hiGiU:btpe124) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1698)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks like you missed the second suggestion. Not terribly important, `In` works too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/db.cc, line 1694 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlV5_cXv7iJX0XPtfM:-Krls5u7-5kJ0skpPPaz:b-8v45os) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1694)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`encodedLocalStoreSyncKey == kKeyLocalStoreSync`. Or it should. 
</blockquote></details>

Of course.  Bah.

---

*[c-deps/libroach/db.cc, line 1698 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-KrlRm-PcFAWodJEvcn0:-KrlrmzIrnn4dNzpG59W:b-9h0ncv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a00b2af92a8a6510b8533d42fc74308f06a8b178/c-deps/libroach/db.cc#L1698)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Looks like you missed the second suggestion. Not terribly important, `In` works too.
</blockquote></details>

Bah, I did.  Fixed and capitalized Windows too.  (also fixed it in the commit notes).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-:-Krlt0kQK3MDmjDqtXb-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
After the latest revision, this is still working well on Windows.

---

Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17735#-:-KrlsZSMqjpI2q7aGUN5:bnsalj9)*
<!-- Sent from Reviewable.io -->
